### ISSUE TITLE
How to map over a list of records with different record entry names

### ISSUE THREAD

**Issue by Michael-Kateregga** [dhall-lang/dhall-haskell#631]
I am trying to do the following:
--  
    let map = https://raw.githubusercontent.com/dhall-lang/Prelude/v2.0.0/package.dhall

in let some_fun = \(x : Text) -> \(y : Text) -> {x = x , y = y}

in let path = "git@github.com:somebranch/"

in let gitbranch = \(x : Text) -> \(y : Text) -> some_fun (path ++ x) y

in let another_fun = \(z : {x : Text, y : Text}) -> gitbranch z.x z.y

in let result = map.`List`.map  {x : Text, y : Text} {x : Text, y : Text}  another_fun [ {a1 = "a", a2  = "a2"}, {b1 = "b", b2  = "b2"}]

in result
--
This can only work if a1, a2 are replaced by x and b1, b2 by y.  However, I want the assignments to have different names i.e. a1, a2, b1 and b2. Is this possible? I suspect I need some kind of polymorphic function for mapping over my list of records?

**Comment by jneira** [comment]
Hi, a common solution (discussed in #234 f.e.) that maybe fits your need is use assoc lists instead maps: `[ { key: Text, value: Text } ]` but you lose the type safety if you know that you have a known set of possible record types.
Not sure if a polymorphic function could help cause all elements of a list must have the same type. Of course you can have a function to each type or record but it could be cumbersome to maintain.

**Comment by Michael-Kateregga** [comment]
Sorry I figured that all list entries must have the same type.

**Comment by Michael-Kateregga** [comment]
@jneira thanks - yes a function to each type is where I started and was too much. Let me try out your suggestion.

**Comment by Gabriel439** [comment]
@Michael-Kateregga: Note that you can't even stick the two records in a list if they have different field names

However, even if you could store them in the same list, it's not clear how `another_fun` would know which record fields it was supposed to operate on.  In other words, how does it know that the `a1` field is supposed to be the `x` field and that the `a2` field is supposed to be the `y` field?

**Comment by Michael-Kateregga** [comment]
@Gabriel439  yes you are absolutely right, the order in a record doesn't indeed matter. By the way does dhall support pattern matching?

**Comment by Michael-Kateregga** [comment]
@Gabriel439 and @jneira 
How do I delete 
(i) a field from a record?
(ii) an entry from a list?

**Comment by Gabriel439** [comment]
@Michael-Kateregga: Dhall supports `merge` (a way to process the result of a union), but the syntax doesn't look like pattern matching.  See: https://hackage.haskell.org/package/dhall-1.17.0/docs/Dhall-Tutorial.html#g:12

There is no way to delete a field from a record.  There also is no (easy) way to delete an element from a list.  In general, the idiom is to emphasize "additive" programming, where if you don't want an element in a record or list you omit it from the start.  However, if you give an example of your use case I might be able to see if there is another way to accomplish what you want.

**Comment by Michael-Kateregga** [comment]
Thank you @Gabriel439

**Comment by ocharles** [comment]
Well, you can always "delete" a field by building a new record that omits
that field. Eg, \( a : { x : A, y : Y }) -> { y = a.y }

**Comment by Gabriel439** [comment]
You can also use the support for projecting a subset of fields, too.  For example:

```haskell
    let point = { x = 1.0, y = 0.0, z = 2.0 }

in  point.{ x, y }
```

**Comment by Michael-Kateregga** [comment]
@Gabriel439  and @ocharles - yes defining a function that builds a new record as a way of deleting functions works fine, thank you.

One more question, is it possible to change/replace the field name of a record e.g. {firstName = "Michael", lastName = "Kateregga"} to {middleName = "Michael", lastName = "Kateregga"}?

**Comment by jneira** [comment]
Maybe it only would be possible wih dependent types, cause that function would map values to types, right?. 

Using association list that would be possible but i guess you want to have the record type.

Actually i think @f-f did [something similar](https://github.com/dhall-lang/dhall-lang/issues/169#issuecomment-414916674) in two steps (using "macros" :stuck_out_tongue_closed_eyes: ):
1.- Create a dhall file that outputs another dhall file with the record(s) definition(s)
2.- Use it in the "runtime" dhall main file
